### DISCUSSION TITLE
Is there more documentation, an example, more explanation about this module?

### DISCUSSION THREAD

**Discussion by jokoon** [GameNetworking/network_synchronizer#109] [General]
I am looking for a way to have prediction/rollback/reconciliation for a FPS game with Godot.

I found this repo, it is written in C++, which is quite great, and it also has many maintainers

Would it be possible to have a longer readme, maybe some doc?

Is that module being used in a game?

**Comment by AndreaCatania** [comment]
Hello, I'm planning to add some documentation on how to use this module soon. In the meanwhile you can find some info on the [wiki](https://github.com/GameNetworking/network_synchronizer/wiki) and an [old example project](https://github.com/GameNetworking/example-project).

Most of the documentation can be found inside the C++ classes: https://github.com/GameNetworking/network_synchronizer/blob/main/scene_synchronizer.h#L48

> Is that module being used in a game?

Yes this module is being used by: With Me, [Halcyon zero (Pahdo labs)](https://www.halcyon-zero.com/), and [The Mirror](https://twitter.com/themirrorgdp/status/1718019599361323023).
Some other small games, made with godot 3.x, have been using this too.

**Reply by jokoon** [reply]
What is really missing is the build script, which I think is a SCons script.

EDIT: oh sorry it's a module, so it's meant to rebuild godot

**Reply by AndreaCatania** [reply]
Yes, that's correct. Also, I want to notify you that I added some doc here: https://github.com/GameNetworking/network_synchronizer/blob/main/doc_classes/GdSceneSynchronizer.xml#L6

Let me know if you have any feedback / question.

**Comment by jokoon** [comment]
Ah thanks!

Do you think that addon is suitable for a FPS shooter?

**Reply by AndreaCatania** [reply]
Yes, this is designed to do exactly that.

**Comment by jokoon** [comment]
I was discussing on the GDextension channel on discord, and I was asking
why you made a C++ module that requires recompiling godot, instead of
releasing it as a GDextension, which slowed me down a bit. From what other
people said, it seems that a GDextension might incur some performance
penalty, which is the reason you went to the module route.

I'm in the early stage of making the game (something close to counter
strike, low poly, not TOO ambitious), so implementing a proper netcode is
not an emergency, as compiling godot is a bit long.

Is performance the reason you made a module instead of a GDextension? They
also said there is little difference between a module and a GDextension.

Anyway, thanks for writing that doc, that's helpful, I hope I will reach
the day I will use it, proper netcode is difficult to approach. The
Gambetta article was also better than GafferOnGames, it was more detailed
and covered those topics more thoroughly.

**Reply by AndreaCatania** [reply]
Glad you find the docs useful, I'll add more in future!

Regarding your GdExtension question: this library was written when only GDNative existed, and back then it was slower and quite difficult to use.
So, I wanted to stick with a much easier to maintain solution, and that was creating a module.

I don't have plan to convert or add the GDExtension support, but if someone implements it I'm ok with that.

Let me highlight that whether you chose to use it, you will have to build godot only one time and not each time you add a new feature. Compiling godot takes around 10/15 minutes, and less than a minute on future compilations, on an ok computer.

Cheers
